src/persist-client/src/examples/persistent_source_example.rs

```diff
@@ -0,0 +1,989 @@
+    //         source2,
+    //         timestamper,
+    //         batch_size,
+    //         data_write,
+    //     )
+    //     .await
+    //     {
+    //         error!("error in source: {:?}", e);
+    //     }
+    // });
+
+    source_pipeline1.await?;
+    // source_pipeline2.await?;
+
+    reader_pipeline1.await?;
+
+    Ok(())
+}
+
+async fn new_mem_persist() -> Result<Client, LocationError> {
+    let location = Location {
+        bucket: "unused".into(),
+        prefix: "unused".into(),
+    };
+    Client::new(NO_TIMEOUT, location, None).await
+}
+
+const NO_TIMEOUT: Duration = Duration::from_secs(1_000_000);
+
+mod api {
+    use std::fmt::Debug;
+
+    use async_trait::async_trait;
+    use timely::progress::Antichain;
+
+    #[derive(Debug)]
+    pub struct Batch<D, T> {
+        pub updates: Vec<(D, T)>,
+
+        /// The lowest timestamp(s) of updates in this batch.
+        pub lower: Antichain<T>,
+
+        /// The highest timestamp(s) of updates in this batch. This is different from a
+        /// _differential upper_, which is exclusive and denotes the frontier at which new updates
+        /// would happen. For reclocking puposes (finding/minting a matching timestamp binding) we
+        /// are interested in the largest source timestamps (think Kafka partition/offset pairs)
+        /// for which we need to find a binding.
+        pub inclusive_upper: Antichain<T>,
+
+        /// The upper frontier (times at which this source might still produce updates) at the time
+        /// of creating this batch.
+        ///
+        /// Note: Intuitively, this might be just `inclusive_upper + 1` (+1 on each element of the
+        /// Antichain). But it might be different, and computing `t + 1` on a timestamp might not
+        /// be straigtforward, i.e. the `Timestamp` trait doesn't support it.
+        pub upper: Antichain<T>,
+    }
+
+    /// `FT` is `FromTime` and `TT` is `ToTime`.
+    ///
+    /// We carry information about the source frontier inline with batches, so that we can know the
+    /// frontier in the task that does emission. TODO(aljoscha): Think about if this makes sense.
+    #[derive(Debug)]
+    pub struct TimestampedBatch<D, FT, TT> {
+        pub updates: Vec<(D, TT)>,
+
+        pub inclusive_upper: Antichain<TT>,
+
+        /// The upper frontier (times at which this source might still produce updates) at the time
+        /// of creating this batch.
+        pub source_upper: Antichain<FT>,
+    }
+
+    #[async_trait]
+    pub trait Source: Clone {
+        type SourceTimestamp: timely::PartialOrder + PartialOrd + Debug;
+        type D: Debug;
+
+        /// Reads up to `max_num_records` updates from the source. Returns `None` if no data is
+        /// available.
+        ///
+        /// We could think about returning an empty batch with the same lower/inclusive_upper as
+        /// last time. Or we could think about this method just blocking forever, but it doesn't
```

Comment:
the important thing to note is there are 2 ways to "block forever":
- Actually blocking the thread, by doing busy-waiting or doing "blocking io"
- "blocking the async fn from making progress", by yielding forever

i think the rust async world leans on only using the word "blocking" for the former to avoid confusion